larpandoracontent/LArThreeDReco/LArCosmicRay/TwoViewCosmicRayRemovalTool.cc

```diff
@@ -0,0 +1,560 @@
+/**
+ *  @file   larpandoracontent/LArThreeDReco/LArTransverseTrackMatching/TwoViewCosmicRayRemovalTool.cc
+ *
+ *  @brief  Implementation of the two view cosmic removal tool class.
+ *
+ *  $Log: $
+ */
+
+#include "Pandora/AlgorithmHeaders.h"
+#include "larpandoracontent/LArThreeDReco/LArCosmicRay/TwoViewCosmicRayRemovalTool.h"
+
+#include "larpandoracontent/LArHelpers/LArClusterHelper.h"
+#include "larpandoracontent/LArHelpers/LArGeometryHelper.h"
+#include "larpandoracontent/LArHelpers/LArMuonLeadingHelper.h"
+#include "larpandoracontent/LArObjects/LArTwoDSlidingFitResult.h"
+#include "larpandoracontent/LArHelpers/LArPfoHelper.h"
+
+using namespace pandora;
+
+namespace lar_content
+{
+
+TwoViewCosmicRayRemovalTool::TwoViewCosmicRayRemovalTool() :
+    m_minSeparation(2.f),
+    m_slidingFitWindow(10000),
+    m_distanceToLine(0.5f),
+    m_minContaminationLength(3.f),
+    m_maxDistanceToHit(1.f),
+    m_minRemnantClusterSize(3),
+    m_maxDistanceToTrack(2.f)
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool TwoViewCosmicRayRemovalTool::Run(TwoViewDeltaRayMatchingAlgorithm *const pAlgorithm, MatrixType &overlapMatrix)
+{
+    m_pParentAlgorithm = pAlgorithm;
+    
+    if (PandoraContentApi::GetSettings(*m_pParentAlgorithm)->ShouldDisplayAlgorithmInfo())
+       std::cout << "----> Running Algorithm Tool: " << this->GetInstanceName() << ", " << this->GetType() << std::endl;
+    
+    bool changesMade(false);
+
+    ClusterVector sortedKeyClusters;
+    overlapMatrix.GetSortedKeyClusters(sortedKeyClusters);
+
+    ClusterSet usedKeyClusters;
+    for (const Cluster *const pKeyCluster : sortedKeyClusters)
+    {        
+        if (usedKeyClusters.count(pKeyCluster))
+            continue;
+        
+        MatrixType::ElementList elementList;        
+        overlapMatrix.GetConnectedElements(pKeyCluster, true, elementList);
+
+        for (const MatrixType::Element &element : elementList)
+            usedKeyClusters.insert(element.GetCluster1());
+
+        this->ExamineConnectedElements(elementList, changesMade);
+    }
+    
+    return changesMade;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void TwoViewCosmicRayRemovalTool::ExamineConnectedElements(const MatrixType::ElementList &elementList, bool &changesMade) const
```

Comment:
This looks to be the same function as in the CosmicRayRemoval class, so is subject to the same comment as above.